### PR TITLE
Fix coordinator loadStatus performance

### PR DESCRIPTION
**Pull request by jon-wei**
`getLoadStatus()` in `DruidCoordinator` determines how many segments need to be loaded per-datasource by retrieving the list of all segments from metadata, then asks each server for its segment inventory and removes the server segments from the set returned from metadata:

```
      // remove loaded segments
      for (DruidServer druidServer : serverInventoryView.getInventory()) {
        final DruidDataSource loadedView = druidServer.getDataSource(dataSource.getName());
        if (loadedView != null) {
          segments.removeAll(loadedView.getSegments());
        }
      }
```

The current code can perform badly when a server has the entire set of segments for a datasource.

This is the code for `AbstractSet.removeAll()`:
```
    public boolean removeAll(Collection<?> c) {
        Objects.requireNonNull(c);
        boolean modified = false;

        if (size() > c.size()) {
            for (Iterator<?> i = c.iterator(); i.hasNext(); )
                modified |= remove(i.next());
        } else {
            for (Iterator<?> i = iterator(); i.hasNext(); ) {
                if (c.contains(i.next())) {
                    i.remove();
                    modified = true;
                }
            }
        }
        return modified;
    }
```

In that situation, the else block will be used, and this is a problem because `loadedView.getSegments()` is the values collection of a ConcurrentHashMap, and its contains() method does a full traversal of the map:

```
  public Collection<DataSegment> getSegments()
  {
    return Collections.unmodifiableCollection(idToSegmentMap.values());
  }

...
    public static <T> Collection<T> unmodifiableCollection(Collection<? extends T> c) {
        return new UnmodifiableCollection<>(c);
    }
    public boolean contains(Object o)   {return c.contains(o);}
...
    static final class ValuesView<K,V> extends CollectionView<K,V,V>
        implements Collection<V>, java.io.Serializable {

        ValuesView(ConcurrentHashMap<K,V> map) { super(map); }

        public final boolean contains(Object o) {
            return map.containsValue(o);
        }
...
    /**
     * Returns {@code true} if this map maps one or more keys to the
     * specified value. Note: This method may require a full traversal
     * of the map, and is much slower than method {@code containsKey}.
     *
     * @param value value whose presence in this map is to be tested
     * @return {@code true} if this map maps one or more keys to the
     *         specified value
     * @throws NullPointerException if the specified value is null
     */
    public boolean containsValue(Object value) {
        if (value == null)
            throw new NullPointerException();
        Node<K,V>[] t;
        if ((t = table) != null) {
            Traverser<K,V> it = new Traverser<K,V>(t, t.length, 0, t.length);
            for (Node<K,V> p; (p = it.advance()) != null; ) {
                V v;
                if ((v = p.val) == value || (v != null && value.equals(v)))
                    return true;
            }
        }
        return false;
    }
...

```

The following benchmark (included in the PR) shows this behavior:

```
Benchmark                       (serverHasAllSegments)  (totalSegmentsCount)  Mode  Cnt        Score       Error  Units
LoadStatusBenchmark.newVersion                    true                 10000  avgt   50      387.188 ±     4.991  us/op
LoadStatusBenchmark.newVersion                   false                 10000  avgt   50      227.884 ±     2.444  us/op
LoadStatusBenchmark.oldVersion                    true                 10000  avgt   50  1581777.362 ± 38030.835  us/op
LoadStatusBenchmark.oldVersion                   false                 10000  avgt   50      244.024 ±     2.923  us/op
```

### CODE REVIEW

**Comment by gianm** [comment]
> The current code can perform badly when a server has the entire set of segments for a datasource.

This is a pretty neat failure mode!

> 1581777.362 ± 38030.835

And it sounds like "perform badly" is an understatement.

File: server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java [hunk 1]
```diff
@@ -297,7 +297,9 @@ boolean hasLoadPending(final String dataSource)
       for (DruidServer druidServer : serverInventoryView.getInventory()) {
         final DruidDataSource loadedView = druidServer.getDataSource(dataSource.getName());
         if (loadedView != null) {
-          segments.removeAll(loadedView.getSegments());
+          for (DataSegment serverSegment : loadedView.getSegments()) {
```
Comment (gianm) [inline server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java]: This deserves a comment so someone doesn't "simplify" it back into the old code.

Comment (jon-wei) [reply server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java]: Good point, added a comment

**Comment by gianm** [comment]
@jon-wei is there anywhere else using removeAll that looks suspicious? (IMO, given what we see here, anything where the argument is not a Set is suspicious.)

**Comment by jon-wei** [comment]
@gianm 

KafkaSupervisor calls removeAll in checkPendingCompletionTasks, where both collections are ArrayLists, containing task groups 

PendingTaskBasedWorkerProvisioningStrategy and SimpleWorkerProvisioningStrategy have instances where Set.removeAll is called with a List, containing worker IDs

**Pull request merged**